### PR TITLE
2906-2907-2909: Fiji covid vaccination overlays

### PR DESCRIPTION
**Pull request by chris-bes**
### Issues:
https://github.com/beyondessential/tupaia-backlog/issues/2906
https://github.com/beyondessential/tupaia-backlog/issues/2907
https://github.com/beyondessential/tupaia-backlog/issues/2909 

### Changes:
- Fiji covid vaccine map overlays

### CODE REVIEW

File: packages/database/src/migrations/20210602030420-AddFijiCOVIDVaccinatedByFacilityOverlays-modifies-data.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    aggregations: [
+      {
+        type: 'FINAL_EACH_DAY',
+      },
+    ],
+    entityAggregation: {
+      dataSourceEntityType: 'facility',
+    },
+  },
+  measureBuilder: 'sumAllPerOrgUnit',
```
Comment (chris-bes) [inline packages/database/src/migrations/20210602030420-AddFijiCOVIDVaccinatedByFacilityOverlays-modifies-data.js:34]: This migration is for: beyondessential/tupaia-backlog#2906
The overlay is heatmap at facility level. We want to get **latest for each day** for a facility, and sum all of them to get the total number of doses

File: packages/database/src/migrations/20210602065620-AddFijiCOVIDVaccinatedRegionOverlays-modifies-data.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      {
+        type: 'FINAL_EACH_DAY',
+      },
+    ],
+    entityAggregation: {
+      dataSourceEntityType: 'facility',
+      aggregationEntityType,
+    },
+  },
+  measureBuilder: 'sumAllPerOrgUnit',
```
Comment (chris-bes) [inline packages/database/src/migrations/20210602065620-AddFijiCOVIDVaccinatedRegionOverlays-modifies-data.js:37]: This migration is for: beyondessential/tupaia-backlog#2907
The overlay is shaded polygons at district/sub_district level. We want to get latest for each day for a facility, and aggregate (sum) to district/sub_district level

File: packages/database/src/migrations/20210603023401-AddFijiCOVIDVaccinePercentageOfPopulationVaccinatedMapOverlays-modifies-data.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            entityAggregation: {
+              dataSourceEntityType: 'sub_district',
+              aggregationType: denominatorAggregationType,
+              aggregationEntityType: denominatorAggregationEntityType,
+            },
+          },
+        },
+      },
+    },
+    measureBuilder: 'composePercentagePerOrgUnit',
```
Comment (chris-bes) [inline packages/database/src/migrations/20210603023401-AddFijiCOVIDVaccinePercentageOfPopulationVaccinatedMapOverlays-modifies-data.js:65]: This migration is for: beyondessential/tupaia-backlog#2909
The overlay is shaded polygons at district/sub_district level. We want to get **latest for each day for a facility**, and aggregate (sum) to district/sub_district level. And calculate the percentage by dividing the **latest** population

File: packages/database/src/migrations/20210603032718-ReorderFijiCovidVaccineMapOverlays-modifies-data.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  'FJ_COVID_TRACKING_Dose_1_SubDistrict',
+  'FJ_COVID_TRACKING_Dose_1_District',
+  'FJ_COVID_TRACKING_Dose_1_SubDistrict_Percentage_Vaccinated',
+  'FJ_COVID_TRACKING_Dose_1_District_Percentage_Vaccinated',
+  'FJ_COVID_TRACKING_Dose_2_Facility',
+  'FJ_COVID_TRACKING_Dose_2_SubDistrict',
+  'FJ_COVID_TRACKING_Dose_2_District',
+  'FJ_COVID_TRACKING_Dose_2_SubDistrict_Percentage_Vaccinated',
+  'FJ_COVID_TRACKING_Dose_2_District_Percentage_Vaccinated',
+];
```
Comment (chris-bes) [inline packages/database/src/migrations/20210603032718-ReorderFijiCovidVaccineMapOverlays-modifies-data.js:30]: This migration is just to reorder the overlays

File: packages/database/src/migrations/20210602030420-AddFijiCOVIDVaccinatedByFacilityOverlays-modifies-data.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+exports.setup = function (options, seedLink) {
+  dbm = options.dbmigrate;
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+const createMapOverlay = doseNum => ({
+  id: `FJ_COVID_TRACKING_Dose_${doseNum}_Facility`,
+  name: `COVID-19 Vaccine Dose ${doseNum} (Facility)`,
+  userGroup: 'COVID-19',
```
Comment (billli0) [inline packages/database/src/migrations/20210602030420-AddFijiCOVIDVaccinatedByFacilityOverlays-modifies-data.js:22]: ^userGroup should be `Fiji Supply Chain`?

File: packages/database/src/migrations/20210602030420-AddFijiCOVIDVaccinatedByFacilityOverlays-modifies-data.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    },
+  },
+  measureBuilder: 'sumAllPerOrgUnit',
+  presentationOptions: {
+    scaleType: 'neutral',
+    displayType: 'spectrum',
+    scaleBounds: {
+      left: {
+        max: 0,
+        min: 0,
```
Comment (billli0) [inline packages/database/src/migrations/20210602030420-AddFijiCOVIDVaccinatedByFacilityOverlays-modifies-data.js:41]: from card [#2906](https://github.com/beyondessential/tupaia-backlog/issues/2906) it said the lower bound should be 1

File: packages/database/src/migrations/20210602065620-AddFijiCOVIDVaccinatedRegionOverlays-modifies-data.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  presentationOptions: {
+    scaleType: 'neutral',
+    displayType: 'shaded-spectrum',
+    scaleBounds: {
+      left: {
+        max: 0,
+        min: 0,
```
Comment (billli0) [inline packages/database/src/migrations/20210602065620-AddFijiCOVIDVaccinatedRegionOverlays-modifies-data.js:44]: same here.

File: packages/database/src/migrations/20210603023401-AddFijiCOVIDVaccinePercentageOfPopulationVaccinatedMapOverlays-modifies-data.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  const numeratorDataElementCode = doseNum === 1 ? 'COVIDVac4' : 'COVIDVac8';
+  const denominatorAggregationEntityType =
+    aggregationEntityType === 'sub_district' ? undefined : aggregationEntityType;
+  const denominatorAggregationType =
+    aggregationEntityType === 'sub_district' ? undefined : 'SUM_PER_ORG_GROUP';
+
+  return {
+    id: getOverlayId(doseNum, level),
+    name: `% of Population Vaccinated COVID-19 Dose ${doseNum} (${overlayLevelName})`,
+    userGroup: 'COVID-19',
```
Comment (billli0) [inline packages/database/src/migrations/20210603023401-AddFijiCOVIDVaccinePercentageOfPopulationVaccinatedMapOverlays-modifies-data.js:32]: same here

File: packages/database/src/migrations/20210602065620-AddFijiCOVIDVaccinatedRegionOverlays-modifies-data.js [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  type = dbm.dataType;
+  seed = seedLink;
+};
+
+const getOverlayId = (doseNum, level) => `FJ_COVID_TRACKING_Dose_${doseNum}_${level}`;
+
+const createMapOverlay = (doseNum, aggregationEntityType, level, overlayLevelName) => ({
+  id: getOverlayId(doseNum, level),
+  name: `COVID-19 Vaccine Dose ${doseNum} (${overlayLevelName})`,
+  userGroup: 'COVID-19',
```
Comment (billli0) [inline packages/database/src/migrations/20210602065620-AddFijiCOVIDVaccinatedRegionOverlays-modifies-data.js:24]: same here

**Review by billli0 — APPROVED** [review]
Look goods to me overall! @chris-bes Would like to check with you for a few details. Will do a pre-approve.

File: packages/database/src/migrations/20210603023401-AddFijiCOVIDVaccinePercentageOfPopulationVaccinatedMapOverlays-modifies-data.js [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+const getOverlayId = (doseNum, level) =>
+  `FJ_COVID_TRACKING_Dose_${doseNum}_${level}_Percentage_Vaccinated`;
+
+const createMapOverlay = (doseNum, aggregationEntityType, level, overlayLevelName) => {
+  const numeratorDataElementCode = doseNum === 1 ? 'COVIDVac4' : 'COVIDVac8';
+  const denominatorAggregationEntityType =
+    aggregationEntityType === 'sub_district' ? undefined : aggregationEntityType;
+  const denominatorAggregationType =
+    aggregationEntityType === 'sub_district' ? undefined : 'SUM_PER_ORG_GROUP';
```
Comment (billli0) [inline packages/database/src/migrations/20210603023401-AddFijiCOVIDVaccinePercentageOfPopulationVaccinatedMapOverlays-modifies-data.js:27]: having tried to put undefined as value for these configs, if that doesn't throw errors that should be fine.

**Comment by chris-bes** [comment]
@billli0 Have addressed your comments!

**Pull request merged**